pkg/ioutils/readers.go

```diff
@@ -14,6 +14,34 @@ import (
 
 var rndSrc = random.NewSource()
 
+// ReadCloserWithError is used to wrap a reader but also be able to close the
+// reader given a specific error. Successive calls to Read() will return the
+// given error.
+type ReadCloserWithError struct {
+	io.Reader
+	err error
```

Comment:
You're right, in general this would be bad, but I was only thinking about in the the use case of the layer downloads. It would allow one read to continue but the next read would fail. See my comment in the main thread about using a Pipe.